core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/aggregation/aggregators/TwoStepNumbersAggregator.kt

```diff
@@ -0,0 +1,167 @@
+                emptyInput = col.size() == 0,
+            ) ?: value::class.starProjectedType // heavy fallback type calculation
+
+            value to type
+        }.unzip()
+
+        return aggregateCalculatingType(
+            values = values,
+            valueTypes = types.toSet(),
+        )
+    }
+
+    /**
+     * Base function of [Aggregator].
+     *
+     * Aggregates the given values, taking [type] into account, and computes a single resulting value.
+     *
+     * Uses [aggregator] to compute the result.
+     *
+     * This function is modified to call [aggregateCalculatingType] when it encounters mixed number types.
+     * This is not optimal and should be avoided by calling [aggregateCalculatingType] with known number types directly.
+     *
+     * When the exact [type] is unknown, use [aggregateCalculatingType].
+     */
+    override fun aggregate(values: Iterable<Number>, type: KType): Return? {
+        require(type.isSubtypeOf(typeOf<Number?>())) {
+            "${TwoStepNumbersAggregator::class.simpleName}: Type $type is not a subtype of Number?"
+        }
+
+        return when (type.withNullability(false)) {
+            // If the type is not a specific number, but rather a mixed Number, we unify the types first.
+            // This is heavy and could be avoided by calling aggregate with a specific number type
+            // or calling aggregateCalculatingType with all known number types
+            typeOf<Number>() -> aggregateCalculatingType(values)
+
+            // Nothing can occur when values are empty
+            nothingType -> super.aggregate(values, type)
+
+            !in primitiveNumberTypes -> throw IllegalArgumentException(
+                "Cannot calculate $name of ${renderType(type)}, only primitive numbers are supported.",
+            )
+
+            else -> super.aggregate(values, type)
+        }
+    }
+
+    /**
+     * Special case of [aggregate] with [Iterable] that calculates the [unified number type][UnifyingNumbers]
+     * of the values at runtime and converts all numbers to this type before aggregating.
+     * Without [valueTypes], this is a heavy operation and should be avoided when possible.
+     *
+     * @param values The numbers to be aggregated.
+     * @param valueTypes The types of the numbers.
+     *   If provided, this can be used to avoid calculating the types of [values] at runtime with reflection.
+     *   It should contain all types of [values].
+     *   If `null`, the types of [values] will be calculated at runtime (heavy!).
+     */
+    @Suppress("UNCHECKED_CAST")
+    override fun aggregateCalculatingType(values: Iterable<Number>, valueTypes: Set<KType>?): Return? {
+        val valueTypes = valueTypes ?: values.types()
+        val commonType = valueTypes
+            .unifiedNumberType(PRIMITIVES_ONLY)
+            .withNullability(false)
+
+        if (commonType == typeOf<Double>() && (typeOf<ULong>() in valueTypes || typeOf<Long>() in valueTypes)) {
```

Comment:
For Float and Long, the precision loss can still occur but to a lesser extent. 

Float has a lower precision compared to Double, and converting Long to Float can also result in precision loss, although less severe compared to Double.

Why do we avoid this in your solution for Floats?

Reply:
Unification only happens upwards, so Floats become Doubles but not the other way around. Longs will never become Floats. With PRIMITIVES_ONLY, Longs can however become Double now with a slight loss in precision (we would use BigDecimal earlier), hence the warning.